Initial E2E system tests (`set_code, authorize_upgrade`)

karolk91 opened the pull request:
Relates to: https://github.com/open-web3-stack/polkadot-ecosystem-tests/issues/298

Will close #456 .

This initial implementation provides 

Scenarios for Polkadot and Kusama relays. These scenearios utilize local pallet Scheduler to execute upgrade related calls as Root:
- call `set_code` with currently used WASM expecting error to happen (this path is checking eg `specVersion` from WASM and will not allow upgrade to pass if it hasn't been incremented)
- call `set_code_without_checks` with currently used WASM expecting success, basically upgrading to the same wasm but version checks are skipped on runtime level thus allowing to pass
- two step approach to upgrade -> `authorize_upgrade` + `apply_authorized_upgrade` with currently used WASM expecting error to happen
- two step approach to upgrade -> `authorize_upgrade_without_checks` + `apply_authorized_upgrade` with currently used WASM expecting success
- check that subsequent `authorize_upgrade` and `authorize_upgrade_without_checks` can override authorized hash


Scenarios for Polkadot/Kusama system chains (AH, BH, People, Coretime, Collectives, Encointer). These scenarios utilize Relay chain's pallet Scheduler to send XCM Transact go gain root on destination chain:
- two step approach to upgrade -> `authorize_upgrade` + `apply_authorized_upgrade` with currently used WASM expecting error to happen
- two step approach to upgrade -> `authorize_upgrade_without_checks` + `apply_authorized_upgrade` with currently used WASM expecting success
- check that subsequent `authorize_upgrade` and `authorize_upgrade_without_checks` can override authorized hash

rockbmb commented on packages/shared/src/system.ts:
```diff
@@ -0,0 +1,250 @@
+    },
+  ])
+
+  const authroizedUpgrade = (await client.api.query.system.authorizedUpgrade()).value
```
```suggestion
  const authorizedUpgrade = (await client.api.query.system.authorizedUpgrade()).value
```

rockbmb commented on packages/shared/src/system.ts:
```diff
@@ -0,0 +1,250 @@
+      },
+    ])
+
+    const authroizedUpgrade = (await client.api.query.system.authorizedUpgrade()).value
```
```suggestion
    const authorizedUpgrade = (await client.api.query.system.authorizedUpgrade()).value
```

karolk91 commented on packages/shared/src/system.ts:
```diff
@@ -0,0 +1,250 @@
+type AuthorizeUpgradeFn = (codeHash: string | Uint8Array<ArrayBufferLike>) => SubmittableExtrinsic<'promise'>
+type ExpectedEvents = Parameters<typeof assertExpectedEvents>[1]
+
+async function runAuthorizeUpgradeScenario(
```
@rockbmb , of course, added new part of tests (for system chains) and also introduced some docs in https://github.com/open-web3-stack/polkadot-ecosystem-tests/pull/411/commits/df85f3d06b2c09186e45404d4e07cecb8ec01d58

karolk91 commented on packages/polkadot/src/bridgeHubPolkadot.system.e2e.test.ts:
```diff
@@ -0,0 +1,10 @@
+import { type RelayTestConfig, registerTestTree, systemE2ETestsViaRelay } from '@e2e-test/shared'
+
+const testConfig: RelayTestConfig = {
+  testSuiteName: 'Polkadot Collectives System',
```
You probably reviewed older commit as this has been fixed already

karolk91 commented on packages/shared/src/helpers/index.ts:
```diff
@@ -440,3 +443,109 @@ export interface ParaTestConfig {
+
+type ArgMatcher = unknown | ((actual: unknown) => boolean)
+
+type EventMatchCriteria<T extends AnyTuple = AnyTuple, N = unknown> = {
```
Some docs added in: https://github.com/open-web3-stack/polkadot-ecosystem-tests/pull/411/commits/950bc134ee49e9d749d6b1202502b639162272b4

karolk91 commented on packages/shared/src/system.ts:
```diff
@@ -0,0 +1,569 @@
+
+/**
+ * Runs the authorize upgrade + apply authrozied upgrade scenario
+ * Scenario will fetch WASM from :code storage thus effectively trying to upgrade to the same WASM as currently used
```
Improved docs in https://github.com/open-web3-stack/polkadot-ecosystem-tests/pull/411/commits/950bc134ee49e9d749d6b1202502b639162272b4

rockbmb commented on packages/shared/src/system.ts:
```diff
@@ -0,0 +1,613 @@
+  return runAuthorizeUpgradeScenario(client, {
+    call: client.api.tx.system.authorizeUpgrade,
+    expectedAfterApply: (hash) => [
+      { type: client.api.events.system.RejectedInvalidAuthorizedUpgrade, args: { codeHash: hash } },
```
If it's not too much work, could you also check the `error` field of `RejectedInvalidAuthorizedUpgrade`?

A runtime upgrade to the same WASM failing is expected, but the reasons for the failure should be (and stay) the same.

rockbmb replied:
Indeed - I was unaware that was the issue, as otherwise I'd have pointed you to a few instances of this pattern in PEt.

karolk91 replied:
done in https://github.com/open-web3-stack/polkadot-ecosystem-tests/pull/411/commits/eee23572af1b3c67bf95a18b98f0b4b0d1ad44ab

rockbmb commented on packages/shared/src/system.ts:
```diff
@@ -0,0 +1,613 @@
+) {
+  const authorizeHash = async (someHash) => {
+    const call = params.call(someHash).method
+    await scheduleInlineCallWithOrigin(client, call.toHex(), { system: 'Root' })
```
Every instance of `scheduleInlineCallWithOrigin` needs to be given a block provider, which the test config type has: 

https://github.com/open-web3-stack/polkadot-ecosystem-tests/blob/ae7271ec9f5baa0796d475fc803514f4a5749d51/packages/shared/src/helpers/index.ts#L466-L489

The scheduler in Kusama AH (and soon PAH) keyes its agenda via a nonlocal block provider, which means these tests will fail on KAH.

See other E2E tests for how to do this, e.g. scheduler tests for KAH.

karolk91 replied:
Existing test are actually fine, because the scenarios I introduced are using Relay's scheduler pallet - and that is always a "local" block provider

Before AHM I think parachains didn't have their scheduler (maybe Collectives only?), with AHM we are introducing scheduler pallet to Asset Hubs - so I can possibly create one more scenario specifiacally for post AHM Asset Hub such that it could use its own scheduler to execute the runtime upgrade

karolk91 replied:
`testConfig.blockProvider` and extra scenario introduced in https://github.com/open-web3-stack/polkadot-ecosystem-tests/pull/411/commits/eee23572af1b3c67bf95a18b98f0b4b0d1ad44ab

rockbmb commented:
Will close #456 .

karolk91 commented on packages/kusama/src/assetHubKusama.system.e2e.test.ts:
```diff
@@ -0,0 +1,10 @@
+const testConfig: RelayTestConfig = {
+  testSuiteName: 'Kusama AssetHub System',
+  addressEncoding: 0,
+  blockProvider: 'Local',
```
I have added both scenarios:

- using scheduler on relay --> local block provider (from relay perspective)
- using scheduler on AH (post-AHM) --> NonLocal provider in that case

karolk91 commented:
@rockbmb , few tests for Kusama are failing due to a bug where Preimage pallet is still filtered out post migration and PET scenarios here are extensively using Preimage pallet along with Scheduler to inject stuff for execution

This is affecting current live Kusama Relay. It that has been fixed by https://github.com/polkadot-fellows/runtimes/pull/957 but we will need to wait for Kusama Relay update

karolk91 commented:
- I have commented out failing scenarios for Kusama with TODO to enable them after Kusama 2.0 release (due to preimage pallet bug)
- added new scenarios for Kusama where Kusama Asset Hub is used as the "scheduler / governance" chain for other parachains and relay (so basically its like saying lets use KAH gov to upgrade other chains)
- added commented out post-AHM scenarios for Polkadot - once we merge this to master, I can prepare a PR to target polkadot-ahm branch and we should enable these tests there

rockbmb commented on packages/kusama/src/kusama.system.e2e.test.ts:
```diff
@@ -1,10 +1,20 @@
+
+// registerTestTree(systemE2ETests(kusama, testConfig))
+
+const testConfigForAssetHub: ParaTestConfig = {
```
This was fine as it was, `RelayTestConfig`

rockbmb commented on packages/polkadot/src/polkadot.system.e2e.test.ts:
```diff
@@ -8,3 +8,14 @@ const testConfig: RelayTestConfig = {
+
+// TODO: Uncomment Post-AHM on Polkadot
+
+// const testConfigForAssetHub: ParaTestConfig = {
```
This also needs to changed: `RelayTestConfig`.

I know it's commented out, but still.
